tactic/interactive.lean

```diff
@@ -443,27 +443,15 @@ do e ← to_expr e,
    exp_fields.mmap $ λ ⟨p,n⟩,
      (prod.mk n ∘ to_pexpr) <$> mk_mapp (n.update_prefix p) [none,some e]
 
-/-- `refine_struct { .. }` acts like `refine` but works only with structure instance
-    literals. It creates a goal for each missing field and tags it with the name of the
-    field so that `have_field` can be used to generically refer to the field currently
-    being refined.
-
-    As an example, we can use `refine_struct` to automate the construction semigroup
-    instances:
-    ```
-    refine_struct ( { .. } : semigroup α ),
-    -- case semigroup, mul
-    -- α : Type u,
-    -- ⊢ α → α → α
-
-    -- case semigroup, mul_assoc
-    -- α : Type u,
-    -- ⊢ ∀ (a b c : α), a * b * c = a * (b * c)
-    ```
--/
-meta def refine_struct (e : parse texpr) : tactic unit :=
-do    str ← e.get_structure_instance_info,
-      tgt ← target,
+meta def collect_struct : pexpr → state_t (list $ expr×structure_instance_info) tactic pexpr | e :=
```

Comment:
I'm curious about this style. Why

```lean
meta def collect_struct : pexpr → state_t (list $ expr×structure_instance_info) tactic pexpr | e :=
```

and not

```lean
meta def collect_struct (e : pexpr) : state_t (list $ expr×structure_instance_info) tactic pexpr :=
```

?

Reply:
That's because `collect_struct` is recursive. The arguments left of `:` remain constant through recursive calls

Reply:
Oh, right, didn't look into the function. 😄

Reply:
👍 

As an aside, in non-meta code, this is the kind of recursive function would require general recursion. Termination would be hard to prove otherwise.